audio-device-client/explainer.md

```diff
@@ -0,0 +1,167 @@
+source code ensures the identical sonic result across various platforms.
+(e.g. encoding/decoding, signal processing or an entire audio application)
+
+However, this convoluted workaround only solves some of problems that this breed
+of audio-centric applications faces. It is still locked by the graph render
+quantum size (128 sample-frames), and there is no unified API that configures
+essential properties of audio rendering mechanism (such as I/O device selection,
+multi-channel I/O support, configurable sample rate and more).
+
+The slow adoption of Web Audio API from pro-audio or game industry is the clear
+evidence of why this problem is important to solve.
+
+
+## Key Features
+
+- Provides a dedicated global scope on a separate thread
+  (real-time/high-priority when permitted) for audio processing purpose
+- Selecting audio I/O devices can be done via `MediaTrackConstraints` pattern
+- Supports variable callback buffer size (as opposed to 128 sample-frames
+  limitation of Web Audio API)
+- Provides implicit sample rate conversion when needed
+- Serves I/O audio data from hardware in a single callback function
+- Provides an AudioContext instance when requested
+- HTTPS only and follows Autoplay policy
+
+
+## Non-goals
+
+- Is NOT to replace Web Audio API
+- Contains no media encoding/decoding service
+
+
+## Potential Use Cases
+
+- A __teleconference app__ with custom audio processing and direct hardware
+  access (e.g. complex echo cancellation, source separation, audio
+  spatialization, and auditory scene analysis/music information retrieval)
+- Porting __existing audio engines__ to the web platform with the minimum
+  engineering cost.
+- A __hybrid audio application__ that uses Web Audio API's graph system
+  (processing and synthesis) and customized audio hardware configuration.
+
+
+## How Audio Device Client Works
+
+![Audio Device Client](https://github.com/WebAudio/web-audio-cg/blob/master/audio-device-client/images/img-2-audio-device-client.png "Audio Device Client")
+
+As discussed, the Audio Device Client will be an intermediate layer between Web
+Audio API and actual audio devices used by browser's audio service. When it is
+instantiated, UA will configures hardware accordingly and sets up a global scope
+with a audio rendering thread.
+
+For querying constraints, a pattern of `mediaDevices.enumerateDevices()` can
+be used. A device client constructor takes a constraint dictionary and generates
+an instance when the query is acceptable by UA. Otherwise, the promise will be
+rejected.
+
+```js
+/* async scope */
+
+const devices = await navigator.mediaDevices.enumerateDevices();
+
+// An imaginary helper function for selecting device IDs for I/O.
+const audioDeviceIds = getMyAudioDeviceIds(devices);
+
+const clientConstraints = {
+  inputDeviceId: audioDeviceIds.input,
+  outputDeviceId: audioDeviceIds.output,
+  sampleRate: 8000,
+  callbackBufferSize: 512,
```

Comment:
Is `callbackBufferSize` optional or required?  If required, that seems hard to get a match on Android.

Reply:
By design, UA can reject the request if you don't get a match.